crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -43,154 +42,464 @@ mod validate_op_tests;
+        // rejected and don't have dependencies.
+        let dependency = get_dependency(op_type, &action);
+
+        // Note that this is async only because of the signature checks done during countersigning.
+        // In most cases this will be a fast synchronous call.
+        let r = validate_op(&op, &dna_def, current_validation_dependencies.clone()).await;
+
+        match r {
+            Ok(outcome) => validation_outcomes.push((op_hash, outcome, dependency)),
+            Err(e) => {
+                tracing::error!(error = ?e, "Error validating op");
+            }
+        }
+    }
+
+    let summary: OutcomeSummary = workspace
+        .dht_db
+        .write_async(move |txn| {
+            let mut summary = OutcomeSummary::default();
+            for (op_hash, outcome, dependency) in validation_outcomes {
+                match outcome {
+                    Outcome::Accepted => {
+                        summary.accepted += 1;
+                        put_validation_limbo(txn, &op_hash, ValidationLimboStatus::SysValidated)?;
+                    }
+                    Outcome::MissingDhtDep(missing_dep) => {
+                        summary.missing += 1;
+                        let status = ValidationLimboStatus::AwaitingSysDeps(missing_dep);
+                        put_validation_limbo(txn, &op_hash, status)?;
+                    }
+                    Outcome::Rejected => {
+                        summary.rejected += 1;
+                        if let Dependency::Null = dependency {
+                            put_integrated(txn, &op_hash, ValidationStatus::Rejected)?;
+                        } else {
+                            put_integration_limbo(txn, &op_hash, ValidationStatus::Rejected)?;
+                        }
+                    }
                 }
-                .boxed()
             }
-        },
-        |batch| {
-            let space = space.clone();
+            WorkflowResult::Ok(summary)
+        })
+        .await?;
+
+    tracing::debug!(
+        ?summary,
+        ?num_ops_to_validate,
+        "Finished sys validation workflow"
+    );
+
+    Ok(summary)
+}
+
+fn get_dependency_hashes_from_actions<A>(actions: A) -> Vec<ActionHash>
+where
+    A: Iterator<Item = Action>,
+{
+    actions
+        .flat_map(|action| {
+            vec![
+                match action.prev_action().cloned() {
+                    None => None,
+                    hash => hash,
+                },
+                match action {
+                    Action::Update(action) => Some(action.original_action_address),
+                    Action::Delete(action) => Some(action.deletes_address),
+                    Action::DeleteLink(action) => Some(action.link_add_address),
+                    _ => None,
+                },
+            ]
+            .into_iter()
+            .flatten()
+        })
+        .collect()
+}
+
+async fn fetch_previous_actions<A, C>(
```

Comment:
This needs to exist for inline validation only, but it does at least share code with the main workflow by using the `get_dependency_hashes_from_actions` function